src/textual/widgets/_data_table.py

```diff
@@ -322,82 +427,233 @@ def __init__(
         classes: str | None = None,
     ) -> None:
         super().__init__(name=name, id=id, classes=classes)
+        self.data: dict[RowKey, dict[ColumnKey, CellType]] = {}
+        """Contains the cells of the table, indexed by row key and column key.
+        The final positioning of a cell on screen cannot be determined solely by this
+        structure. Instead, we must check _row_locations and _column_locations to find
+        where each cell currently resides in space."""
+
+        self.columns: dict[ColumnKey, Column] = {}
+        """Metadata about the columns of the table, indexed by their key."""
+        self.rows: dict[RowKey, Row] = {}
+        """Metadata about the rows of the table, indexed by their key."""
+
+        # Keep tracking of key -> index for rows/cols. These allow us to retrieve,
+        # given a row or column key, the index that row or column is currently
+        # present at, and mean that rows and columns are location independent - they
+        # can move around without requiring us to modify the underlying data.
+        self._row_locations: TwoWayDict[RowKey, int] = TwoWayDict({})
+        """Maps row keys to row indices which represent row order."""
+        self._column_locations: TwoWayDict[ColumnKey, int] = TwoWayDict({})
+        """Maps column keys to column indices which represent column order."""
 
-        self.columns: list[Column] = []
-        self.rows: dict[int, Row] = {}
-        self.data: dict[int, list[CellType]] = {}
-        self.row_count = 0
-        self._y_offsets: list[tuple[int, int]] = []
         self._row_render_cache: LRUCache[
-            tuple[int, int, Style, int, int], tuple[SegmentLines, SegmentLines]
-        ]
-        self._row_render_cache = LRUCache(1000)
-        self._cell_render_cache: LRUCache[
-            tuple[int, int, Style, bool, bool], SegmentLines
-        ]
-        self._cell_render_cache = LRUCache(10000)
-        self._line_cache: LRUCache[tuple[int, int, int, int, int, int, Style], Strip]
-        self._line_cache = LRUCache(1000)
+            RowCacheKey, tuple[SegmentLines, SegmentLines]
+        ] = LRUCache(1000)
+        """For each row (a row can have a height of multiple lines), we maintain a
+        cache of the fixed and scrollable lines within that row to minimise how often
+        we need to re-render it. """
```

Comment:
May want reformatting to confirm with [our docstring guidelines](https://github.com/Textualize/textual/wiki/Documentation-standards#class).

Reply:
I'm not going to bother doing this for the internal things. They're mostly just notes for us since the class is quite complex. I'll update the public facing stuff though.